### ISSUE TITLE
FP16 inference latency

### ISSUE THREAD

**Issue by psasidharan** [pytorch/pytorch#44504]
[latency_test.py.txt](https://github.com/pytorch/pytorch/files/5204664/latency_test.py.txt)
## 🐛 Bug

Observing a latency variation of 3-4ms on half precision inference runs when we run inference intermittently (Sleeping 100ms between inference runs). We don't see this latency drop if inference is done from a tight loop. The issue can be reproduced with the attached python code (latency_test.py.txt). The attached python code uses resnet50 from torchvision library to reproduce the issue. 

## To Reproduce

1. Rename latency_test.py.txt to latency_test.py
2. python latency_test.py
3. Observe the prints
                Resnet50 FP16 tight loop:22.886486053466797 ms
                Resnet50 FP16 with delay:25.635948181152344 ms
                Resnet50 FP32 tight loop:46.42632484436035 ms
                Resnet50 FP32 with delay:45.696444511413574 ms

## Expected behavior

Latency variation shouldn't be this high for intermittent runs. Don't see this much variation in FP32 inference.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.5.0+cu101
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version: 
 - CUDA/cuDNN version: 10.1, 7603
 - GPU models and configuration: GeForce RTX 2080 Ti
 - Any other relevant information:

**Comment by HamidShojanazeri** [comment]
Hi @psasidharan 
I was not able to reproduce the issue using the code provided. Although going through the code, I believe time.time() is not monotonic and might not be very helpful in measuring the time intervals. I have tried the code using time.perf_counter() and time.monotinic() which gives consistent results and show no latency more than the delay(sleep for 100ms) in the loops, here is the output log of using these functions
[latency.log](https://github.com/pytorch/pytorch/files/5205286/latency.log)
. It would be great if you could verify using these functions on your side to see if that helps. Also, this [stackoverflow answer](https://stackoverflow.com/questions/25785243/understanding-time-perf-counter-and-time-process-time#:~:text=process_time%20is%20a%20value%20which,is%20running%20on%20the%20CPU) provides a good explanation about their differences. Please let me know if that helps or still facing the same issue.

**Comment by colesbury** [comment]
I was also not able to reproduce this issue. WIth Quadro GP100, CUDA 10.0 PyTorch 1.7.0a0+c44b1de I got:

```
Resnet50 FP16 tight loop:49.96095657348633 ms
Resnet50 FP16 without tight loop:49.18293476104736 ms
Resnet50 FP32 tight loop:59.971885681152344 ms
Resnet50 FP32 without tight loop:60.494399070739746 ms
```

I also think `time.time()` is probably OK for these examples; in theory it could be inaccurate due to time adjustment, but that's unlikely for any single run.

My best guesses are:
1) Your CPU is down clocking on idle. On Linux you can check the scaling governor with:

```
cat /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
```

You can make the CPU less likely to down clock with (but less power efficient):

```
echo performance | sudo tee /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
```

2) You GPU is down clocking on idle. If this is the problem, I don't know how to deal with it.

**Comment by gchanan** [comment]
@colesbury 's suggestion seems like the most likely explanation.

It would also be helpful if you explained your use case.  Is the test after sleeping supposed to be a proxy for a different real-world scenario?

**Comment by psasidharan** [comment]
@gchanan Yes the test after sleeping is a proxt for a real-world scenario. We saw this behavior one our networks (4-6ms latency difference).  @colesbury I have tried setting the governor to performance, but still I see the latency gap even with the perf_counter.

My CPU is Intel(R) Xeon(R) W-2135 CPU @ 3.70GHz & GPU 2080Ti. We don't see this issue in AMD Ryzen with 2080Ti. The whole inference happens on GPU. Any idea why does it vary with CPU?

**Comment by samster25** [comment]
@colesbury @gchanan so I tried the script on many GPUs and theres something interesting here. On a K80, P100, and V100 I get the same exact time within a few microseconds for both loops. However on Turing GPUs (like the T4), I get a gap of milliseconds.
```
PyTorch version: 1.6.0+cu101
cuDNN version: 7603
CUDA version: 10.1
GPU: Tesla T4
Downloading: "https://download.pytorch.org/models/resnet50-19c8e357.pth" to /root/.cache/torch/hub/checkpoints/resnet50-19c8e357.pth
100%
97.8M/97.8M [00:10<00:00, 9.39MB/s]

Resnet50 FP16 tight loop:48.069939613342285 ms
Resnet50 FP16 without tight loop:45.847883224487305 ms
Resnet50 FP32 tight loop:110.69457054138184 ms
Resnet50 FP32 without tight loop:107.71810054779053 ms
```
I put the example up in a [google collab notebook](https://colab.research.google.com/drive/13tp_2XQWs4_vbhYHB1vxAcu5Hg9GmekF?usp=sharing#scrollTo=o7fjG9zokjJp). The notebook will sometimes give you a T4 GPU but you have to constantly terminate the runtime and restart until you get the T4 instance. But when you do, you see the gap.

To terminate the runtime, you can do the following:
`Runtime > Manage Sessions > Terminate`
if it is greyed out, then refreshing the page usually helps.
